vxlan: Add VXLAN properties for EVPN use

varesa opened the pull request:
In order to be able to use VXLAN interfaces as part of an EVPN fabric (separate control plane for VTEP and MAC learning), some expansions are done to the configurable properties:

- Make `base-if` non-mandatory

  EVPN networks are often run on top of a L3 underlay, which means that there are several L3 links to the host where ECMP routing is desired. The VXLAN interface should not be tied to any physical interface (and it's not required since multicast is not used) and the egress interface is determinted through standard routing.

- Allow specifying `local` IP

  Since the VXLAN interface is not tied to any physical interface, it is often desired to source the tunnel from a loopback IP.

- Disable data plane MAC `learning`

  In EVPN MAC learning is handled through a separate control plane like MP-BGP (e.g. FRR as daemon) writing entries to the bridge FDB.

cathay4t commented on rust/src/lib/ifaces/vxlan.rs:
```diff
@@ -48,18 +48,26 @@ impl VxlanInterface {
-        self.vxlan.as_ref().map(|cfg| cfg.base_iface.as_str())
+        self.vxlan
+            .as_ref()
+            .and_then(|cfg| cfg.base_iface.as_ref())
+            .map(|base_iface| base_iface.as_str())
```
How about `as_deref()`?

cathay4t commented on rust/src/lib/ifaces/vxlan.rs:
```diff
@@ -48,18 +48,26 @@ impl VxlanInterface {
 #[non_exhaustive]
 pub struct VxlanConfig {
-    pub base_iface: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub base_iface: Option<String>,
```
This is breaking rust API which we have no plan on releasing nmstate 3.x.

Please use `serde(default)` to store the undefined `base_iface` as emptry.

cathay4t commented on rust/src/lib/ifaces/vxlan.rs:
```diff
@@ -48,18 +48,26 @@ impl VxlanInterface {
     #[serde(deserialize_with = "crate::deserializer::u32_or_string")]
     pub id: u32,
     #[serde(skip_serializing_if = "Option::is_none")]
+    pub learning: Option<bool>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub local: Option<std::net::IpAddr>,
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Please add `///` document here.

varesa replied:
Can you clarify what sort of documentation you want? 

Explaining what these options do in the context of VXLANs in general? Or more about the technical implementation in context of `nmstate`?

There's not much documentation for the existing fields either to align myself with

cathay4t replied:
It is for crate user who look at the document like https://docs.rs/nmstate/latest/nmstate/struct.MacVlanConfig.html#structfield.accept_all_mac .

Yes, I forgot to add document for VXLAN, you may skip this.

cathay4t reviewed (changes requested):
Breaking rust API is not allowed yet. Please use `serde(default)`.

cathay4t commented:
@varesa Can you amend the patch base on above feedback?

varesa commented:
> @varesa Can you amend the patch base on above feedback?

That's my intention, but I've been rather busy the past week (and will be for a while more). Sorry about that

cathay4t commented:
Let me see whether I can allocate sometime to amend your patches to get it merged.

varesa committed:
vxlan: Add VXLAN properties for EVPN use

In order to be able to use VXLAN interfaces as part of an
EVPN fabric (separate control plane for VTEP and MAC learning),
some expansions are done to the configurable properties:

- Make base-if non-mandatory

  EVPN networks are often run on top of a L3 underlay, which
  means that there are several L3 links to the host where
  ECMP routing is desired. The VXLAN interface should not be
  tied to any physical interface (and it's not required since
  multicast is not used) and the egress interface is determined
  through standard routing.

- Allow specifying local IP

  Since the VXLAN interface is not tied to any physical interface,
  it is often desired to source the tunnel from a loopback IP.

- Disable data plane MAC learning

  In EVPN MAC learning is handled through a separate control
  plane like MP-BGP (e.g. FRR as daemon) writing entries to the
  bridge FDB.

Unit test cases and integration test case included.

Signed-off-by: Esa Varemo <git@esav.fi>

cathay4t commented:
I have amend the patch to preserve API stability. Patch manual tested. Merging.
